packages/logger/lib/src/amplify_logger.dart

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:async';
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:ansicolor/ansicolor.dart';
+import 'package:aws_common/aws_common.dart';
+
+import 'log_event.dart';
+import 'log_level.dart';
+
+class AmplifyLogger {
+  static final StreamController<LogEvent> _logStreamController =
+      StreamController<LogEvent>();
+
+  static final Sink<LogEvent> _logSink = _logStreamController.sink;
+
+  /// A map of existing logger instances.
+  static final Map<String, AmplifyLogger> _logsMap = {};
+
+  /// A stream of logs from all logger instances.
+  static final Stream<LogEvent> logStream = _logStreamController.stream;
+
+  /// Default logger instance without a namespace.
+  static get defaultLogger => getInstance();
+
+  LogLevel logLevel = LogLevel.info;
+  static LogLevel globalLogLevel = LogLevel.info;
```

Comment:
I am a little confused about how these two levels are used, and I can imagine customer may be as well.

`globalLogLevel` seems to only be used in the default handler, whereas `logLevel` is used in log method, meaning it gets applied to the custom stream handling and the default handling. Is that what we want?

I think whatever log level filter applies to the default handling should probably be applied to the custom stream handling.

Also, right now if a custom was using the default handler and wanted `debug` logs to print for auth, they would have to set the global level and the auth log level. I think that  having to set two levels will be confusing and cumbersome. I am not suggesting you get rid of the global level, just suggesting the API could be improved.

I am not sure what the ideal API would be here. I would have to put some thought into it. Maybe the `globalLogLevel` is really just a setter that sets all the category levels?

Reply:
@dnys1 has a good solution for this below by using AmplifyLogger as a wrapper on top of the `package:logging`. 

`package:logging` has built in support for hierarchal logs. 

Setting the logLevel of an individual 'package:logging' logger will set the logLevel of all its children loggers, unless those loggers define overrides. 

The main question now is how we would want to expose this behavior. 

In the solution Dillon recommends below, there is no access to the underlying `package_logging' -> `logger`.  But I think if we were expose it we would have a good API for setting global vs local log levels. 

Ex: 
// Assume AmplifyLogger has a getter for `package_logging->logger` 

hierarchicalLoggingEnabled = true; // must be true else all loggers share same log level 

var mainLog = AmplifyLogger(); // main logger 

var authLog = AmplifyLogger.category(Category.Auth) // all category loggers are automatically children loggers of main logger 

mainLog.logger.level = Level.INFO; // sets log level of main and all category loggers

authLog.logger.level = Level.WARNING; // sets log level of just auth and overrides mainLog.logger's log level

Reply:
The above proposal would require importing `amplify_logger` and also the `logging` package from pub.dev 

Instead of directly accessing the `logging` package's logger, we should hide it behind our own API. 

so: 

class AmplifyLogger{

void setLogLevel(LogLevel level){
_logger.level = level
}

Furthermore, the basic AmplifyLogger() implicitly returns the main logger.  Perhaps we can make that more clear by renaming the constructor as AmplifyLogger.main()

What are your thoughts @dnys1 @Jordan-Nelson

Reply:
Lastly, I think we should have a `factory` method for AmplifyLogger that doesn't take `Category` in case the user wants to use our package to define their own logging behavior: 

```
factory AmplifyLogger.category(String namespace) {
    return AmplifyLogger._(Logger('Amplify.${namespace}'));
}
```

Reply:
Yeah we should not expose the dart logging package.   Instead we can just expose our own Amplify logging package and have methods for controlling log level. 

Second point sounds good.  Dillon has a good suggestion for that below whereby our factory method can just take a string: 

AmplifyLogger([String namespace = 'Amplify']) => AmplifyLogger._(Logger(namespace));

I'll update the main PR accordingly